File: lib/journal/Alert.cc
// -*- c++ -*-
//
// michael a.g. aïvázis <michael.aivazis@para-sim.com>
// (c) 1998-2025 all rights reserved


// external support
#include "externals.h"
// forward declarations
#include "forward.h"
// type aliases
#include "api.h"

// the global settings
#include "Chronicler.h"
// message contents
#include "Entry.h"
// the superclass
#include "Renderer.h"
// the declaration
#include "Alert.h"


// metamethods
pyre::journal::Alert::~Alert() {}


// implementation details
void
pyre::journal::Alert::header(
    palette_type & palette, linebuf_type & buffer, const entry_type & entry) const
{
    // get the page
    auto & page = entry.page();
    // if there is nothing to print
    if (page.empty()) {
        // we are done
        return;
    }

    // otherwise, get the notes
    auto & notes = entry.notes();
    // get the severity
    auto severity = notes.at("severity");
    // and the channel name
    auto channel = notes.at("channel");
    // ask the palette for the severity decoration
    auto severityColor = palette[severity];
    // and get the reset sequence
    auto resetColor = palette["reset"];

    // try to lookup the application name
    auto cursor = notes.find("application");
    // if it's there, use it; otherwise use the {severity}
    auto tag = (cursor != notes.end()) ? cursor->second : severity;
    // inject the tag, along with a space
    buffer << severityColor << tag << resetColor << " ";
    // followed by the channel name
    buffer << '(' << severityColor << channel << resetColor << "):" << std::endl;

    // all done
    return;
}


void
pyre::journal::Alert::body(
    palette_type & palette, linebuf_type & buffer, const entry_type & entry) const
{
    // get the page
    auto & page = entry.page();
    // if the page is empty
    if (page.empty()) {
        // do nothing
        return;
    }

    // otherwise, get the notes
    auto & notes = entry.notes();
    // get the severity
    auto severity = notes.at("severity");
    // the reset sequence
    auto resetColor = palette["reset"];
    // ask the palette for the severity decoration
    auto severityColor = palette[severity];
    // and the body color
    auto bodyColor = palette["body"];

    // go through the lines in the page
    for (auto line : page) {
        // and render each one
        buffer << severityColor << _bodyMarker << bodyColor << line << resetColor << std::endl;
    }

    return;
}


void
pyre::journal::Alert::footer(
    palette_type & palette, linebuf_type & buffer, const entry_type & entry) const
{
    // get the page
    auto & page = entry.page();
    // if there is nothing to print
    if (page.empty()) {
        // we are done
        return;
    }
    // get the notes
    auto & notes = entry.notes();
    // get the severity
    auto severity = notes.at("severity");
    // and the channel name
    auto channel = notes.at("channel");
    // the reset sequence
    auto resetColor = palette["reset"];
    // ask the palette for the severity decoration
    auto severityColor = palette[severity];

    // at the right decoration level
    if (chronicler_t::decor() > 1) {
        // attempt to get location information
        auto loc = notes.find("filename");
        // if it's there
        if (loc != notes.end()) {
            // extract the filename
            auto filename = loc->second;
            // make some room and turn on location formatting
            buffer << severityColor << _footerMarker << resetColor << "from ";
            // set a maximum length for the rendered filename
            const line_type::size_type maxlen = 80;
            // get the filename size
            auto len = filename.size();
            // so that names that are longer than this maximum
            if (len > maxlen) {
                // get shortened
                buffer << filename.substr(0, maxlen / 4 - 3) << "..."
                       << filename.substr(len - 3 * maxlen / 4);
            } else {
                // otherwise, render the whole name
                buffer << filename;
            }
            // add a spacer
            buffer << ":";

            // attempt to get the line number
            auto & line = notes.at("line");
            // if we know it
            if (!line.empty()) {
                // render it
                buffer << line << ":";
            }
            // save this much
            buffer << std::endl;

            // get the function name
            auto & function = notes.at("function");
            // if we know it
            if (!function.empty()) {
                // render
                buffer
                    // make some room and turn on location formatting
                    << severityColor << _footerMarker
                    << resetColor
                    // the intro
                    << "in '"
                    // the function name
                    << function
                    // terminate and flush
                    << "'" << std::endl;
            }
        }
    }

    // at the right decoration level
    if (chronicler_t::decor() > 2) {
        // render the channel name and severity
        buffer
            // start things off with the marker
            << severityColor << _footerMarker
            << resetColor
            // the channel severity and name
            << "because the " << severityColor << severity << resetColor << " channel '"
            << severityColor << channel << resetColor
            << "' is active"
            // and done
            << std::endl;
    }

    // now for the extra notes, if any
    if (chronicler_t::decor() > 1) {
        // build a set of the keys we have processed already
        std::set<entry_type::key_type> known { "severity", "channel",  "filename",
                                               "line",     "function", "application" };
        // go through the notes
        for (auto & [key, value] : notes) {
            // if the key is in the known pile
            if (known.find(key) != known.end()) {
                // ignore it
                continue;
            }
            // otherwise, render it
            buffer
                // the marker
                << severityColor << _footerMarker
                << resetColor
                // the key
                << severityColor << key << resetColor
                << ": "
                // the value
                << severityColor << value << resetColor << std::endl;
        }
    }

    // all done
    return;
}


// end of file

File: packages/journal/Alert.py
# -*- coding: utf-8 -*-
#
# michael a.g. aïvázis <michael.aivazis@para-sim.com>
# (c) 1998-2023 all rights reserved


# superclass
from .Renderer import Renderer


# the renderer or user-facing essages
class Alert(Renderer):
    """
    The renderer of user-facing messages
    """


    # implementation details
    def header(self, palette, entry):
        """
        Generate the message header
        """
        # get the page
        page = entry.page

        # if there's nothing to do
        if not page:
            # bail
            return

        # get the notes
        notes = entry.notes
        # get the metadata
        channel = notes["channel"]
        severity = notes["severity"]
        application = notes["application"]

        # get the colors
        resetColor = palette["reset"]
        severityColor = palette[severity]

        # tag using either the app name or the channel severity
        tag  = application or severity
        # generate
        buffer = [
            severityColor, tag, resetColor, ":"
            ]
        # if this is an one liner
        if len(page) == 1:
            # add the page contents to the buffer
            buffer += [
                " ",
                page[0],
            ]

        # assemble and push
        yield ''.join(buffer)

        # all done
        return


    def body(self, palette, entry):
        """
        Generate the message body
        """
        # get the page
        page = entry.page

        # if there's nothing to do
        if len(page) < 2:
            # bail
            return

        # get the notes
        notes = entry.notes
        # get the metadata
        severity = notes["severity"]

        # get the colors
        resetColor = palette["reset"]
        bodyColor = palette["body"]
        severityColor = palette[severity]

        # otherwise, go through the page contents
        for line in page:
            # assemble and push
            yield f"{severityColor}{self.bodyMarker}{bodyColor}{line}{resetColor}"

        # all done
        return


    def footer(self, palette, entry):
        """
        Generate the message body
        """
        # get the page
        page = entry.page
        # if there's nothing to do
        if not page:
            # move on
            return

        # get the notes
        notes = entry.notes
        # grab the severity
        severity = notes["severity"]
        # and the name of the channel
        channel = notes["channel"]

        # get the colors
        resetColor = palette["reset"]
        severityColor = palette[severity]

        # set up the marker
        marker = self.footerMarker
        # grab the chronicler
        from . import chronicler

        # if the chronicler decoration level is sufficiently high
        if chronicler.decor > 1:
            # get the name of the file
            filename = notes["filename"]
            # get the line number
            line = str(notes["line"])
            # get the function name
            function = notes["function"]
            # consider it an indication that we have location information
            if filename:
                # lookup the longest filename we are going to print
                maxlen = self.maxlen
                # if the filename is too long
                if len(filename) > maxlen:
                    # compute the leading part
                    filenameLeader = filename[:maxlen//4-3]
                    # the ellipsis
                    ellipsis = "..."
                    # and the trailing part of the filename
                    filenameTrailer = filename[-3*maxlen//4:]
                # otherwise
                else:
                    # the leading part is the whole thing
                    filenameLeader = filename
                    # no ellipsis
                    ellipsis = ""
                    # no trailing part
                    filenameTrailer = ""

                # make a buffer
                buffer = [
                    # a marker
                    severityColor, self.footerMarker, resetColor,
                    # introduce the locator
                    "at ",
                    # the filename
                    f"{filenameLeader}{ellipsis}{filenameTrailer}:",
                    # the line number
                    f"{line}:" if line else "",
                    # the function name
                    f"{function}" if function else ""
                    ]

                # assemble the line and make it available
                yield ''.join(buffer)

        # if the {chronicler} decoration level is sufficiently high
        if chronicler.decor > 2:
            # render the channel info
            buffer = [
                severityColor, marker, resetColor,
                "because the ",
                severityColor, severity, resetColor,
                " channel '",
                severityColor, channel, resetColor,
                "' is active"
                ]
            # assemble and push
            yield ''.join(buffer)

        # now, for the rest of the metadata
        if chronicler.decor > 1:
            # make a pile of the information we have displayed # already
            done = {
                "severity", "channel", "application",
                "filename", "line", "function", "source",
            }
            # go through the metadata
            for key, value in notes.items():
                # if this is something we've dealt with
                if key in done:
                    # skip it
                    continue
                # otherwise, render it
                buffer = [
                    palette[severity], marker, palette["reset"],
                    palette[severity], key, palette["reset"],
                    ": ",
                    palette[severity], value, palette["reset"],
                    ]
                # assemble and push
                yield ''.join(buffer)

        # all done
        return


    # implementation details
    maxlen = 60
    headerMarker = " >> "
    bodyMarker = " -- "
    footerMarker = " .. "


# end of file
